### PR TITLE
GUI: Improve console selection

### PR DESCRIPTION
**Pull request by tunnelsociety**
`_selEnd` could become negative, so the console buffer could be accessed out of bounds.
Dragging outside the console text area resulted in a slightly strange start and/or end selection anchor.
Edit: I wrote before that there was significant increase in CPU usage while dragging beyond top or bottom of console. That was measured wrongly; it's a really minor difference, sorry. Even so, let's avoid the useless computations.

With ASan, try clicking near coordinate (0,0) in the console (within the margin--the black area), or dragging from anywhere out the top-left corner of the window.

The "Improve selection behaviour" commit was tough to explain; here are quick clips demonstrating.
|before|after|
|-|-|
|<video src=https://github.com/user-attachments/assets/cfbcb081-6d66-4967-b348-e68384aa121e>|<video src=https://github.com/user-attachments/assets/871a8be6-291e-438d-b96b-325dc0935f3b>|

(I gave clampSelection() a return value for a convenient `if(clamp())...` usage, but maybe this could be done better?)

```
gui/console.h:185:35: runtime error: index -128 out of bounds for type 'char [131072]'
ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7fcf3646d688 at pc 0x55f8b1c35422 bp 0x7ffdd6882bb0 sp 0x7ffdd6882ba8
READ of size 1 at 0x7fcf3646d688 thread T0
    #0 0x55f8b1c35421 in GUI::ConsoleDialog::drawLine(int) gui/console.cpp:226
    #1 0x55f8b1c371be in GUI::ConsoleDialog::handleMouseMoved(int, int, int) gui/console.cpp:944
    #2 0x55f8b18b4e14 in GUI::GuiManager::processEvent(Common::Event const&, GUI::Dialog*) gui/gui-manager.cpp:869
    #3 0x55f8b18b76cd in GUI::GuiManager::runLoop() gui/gui-manager.cpp:594
    #4 0x55f8b189c43e in GUI::Dialog::runModal() gui/dialog.cpp:78
    #5 0x55f8b1877e25 in GUI::Debugger::enter() gui/debugger.cpp:250
[...]
0x7fcf3646d688 is located 376 bytes to the left of 132200-byte region [0x7fcf3646d800,0x7fcf3648dc68)
allocated by thread T0 here:
    #0 0x7fcf646b94c8 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:95
    #1 0x55f8b1883060 in GUI::Debugger::Debugger() gui/debugger.cpp:57
[...]
```

### CODE REVIEW

**Comment by tunnelsociety** [comment]
That assert is inside a loop. I should move it out.

**Comment by tunnelsociety** [comment]
Actually, why did this happen at all? `buffer()` is `inline char &buffer(int idx) { return _buffer[idx % kBufferSize]; }`; shouldn't this have never happened...?

File: gui/console.cpp [hunk 1]
```diff
@@ -217,12 +218,14 @@ void ConsoleDialog::drawLine(int line) {
 
 	y += line * kConsoleLineHeight;
 
+	int baseIdx = (start + line) * kCharsPerLine;
+	assert(baseIdx >= 0 && baseIdx + limit <= kBufferSize);
```
Comment (bluegr) [inline gui/console.cpp:222]: There's no reason to add an assert in this case. Just return if we are trying to access invalid memory.

Comment (tunnelsociety) [reply gui/console.cpp:222]: Gotcha. I'm hesitant to add silent failure so I'll drop that commit. (I'm just a beginner, though.) Someone else may add that instead.

**Comment by tunnelsociety** [comment]
Trying a new approach; I edited the first comment.

**Comment by bluegr** [comment]
Looks good, thanks for addressing this :)

**Pull request merged**